syntax = "proto3";

import "protos/user.proto";
import "protos/suggestion.proto";

package neva.backend;

service Backend {
  rpc Register(RegisterRequest) returns (GenericReply) {
  }

  rpc Login(LoginRequest) returns (LoginReply) {
  }

  rpc SuggestionItemProposition(SuggestionItemPropositionRequest)
      returns (GenericReply) {
  }

  rpc GetMealSuggestion(GetMealSuggestionRequest)
      returns (GetMealSuggestionReply) {
  }
}

message GenericReply {
}

message RegisterRequest {
  enum AuthenticationType {
    INVALID = 0;
    DEFAULT = 1;
    FACEBOOK = 2;
  };

  User user = 1;
  AuthenticationType authentication_type = 2;
}

message LoginRequest {
  string email = 1;
  string password = 2;
}

message LoginReply {
  // Authentication token to be used in upcoming requests.
  bytes token = 1;
}

message SuggestionItemPropositionRequest {
  // Session key for the user proposing the item.
  bytes token = 1;

  // Holds information on propositions category and name.
  Suggestion suggestion = 2;
}

message GetMealSuggestionRequest {
  // Session key for the user requesting a suggestion.
  bytes token = 1;

  // Category for which the user requests a suggestion.
  Suggestion.SuggestionCategory suggestion_category = 2;
}

message GetMealSuggestionReply {
  // Name of the suggested item.
  string name = 1;
}
